### PR TITLE
Minor bug fixes for course scheduler

### PR DESCRIPTION
**Pull request by charlietharas**
Please don't delete the course-scheduler-rework branch after merging!

### CODE REVIEW

File: src/lib/scheduler.js [hunk 1]
```diff
@@ -119,10 +119,11 @@ const getDefaultSemesterIndex = () => {
   let result = 0;
 
   const now = new Date();
+  console.log(now);
   // Check if Winter (Period 3, above)
   if (
     new Date(DATES.PREREG.WINTER) < now &&
-    now < addDays(new Date(DATES.Winter.END), -7)
+    now < addDays(new Date(DATES.Winter.START))
```
Comment (yechs) [inline src/lib/scheduler.js:126]: Hmmm... I think this is a bit too aggressive.
Maybe change it to 7 days after Winter Study starts (spring registration usually starts the second week of WS).

Also, don't forget to updates the comments on L114 to reflect this logic change.

File: src/components/views/CourseScheduler/Course.jsx [hunk 2]
```diff
@@ -208,19 +208,22 @@ const Course = ({
...
       default:
-        return 1003 + 10 * (course.year % 100);
+        return yearVal + 1;
     }
   };
 
   const bookstoreLink = () => {
     const baseBookstoreLink =
-      "https://www.bkstr.com/williamsstore/follett-discover-view/booklook?shopBy=discoverViewCourse&bookstoreId=506&divisionDisplayName=";
+      "https://www.bkstr.com/williamsstore/course-materials-results?shopBy=course&divisionDisplayName=&bookstoreId=506&programID=912";
```
Comment (yechs) [inline src/components/views/CourseScheduler/Course.jsx:226]: I tried experimenting with the old `follett-discover-view` link, only to realize that they have been completely deprecated.

However, the official Williams catalog is still using them...
For example, try visit [the catalog page for PHIL 225](https://catalog.williams.edu/phil/detail/?strm=1243&cn=225&crsid=020228&req_year=0) and click "View Book Information". The information will be empty.
However, if you use the new `course-materials-result` link ([link](https://www.bkstr.com/williamsstore/course-materials-results?shopBy=course&divisionDisplayName=&departmentDisplayName=PHIL&courseDisplayName=225&sectionDisplayName=T1&programId=912&termId=100082572)), you will find a list of required books.
We should perhaps also report this to whoever is managing the catalog.

File: src/components/views/CourseScheduler/Course.jsx [hunk 3]
```diff
@@ -208,19 +208,22 @@ const Course = ({
...
-        return 1001 + 10 * (course.year % 100);
+        return yearVal - 2; // this does not currently work!
       case "Winter":
-        return 1002 + 10 * (course.year % 100);
+        return yearVal;
       default:
-        return 1003 + 10 * (course.year % 100);
+        return yearVal + 1;
     }
   };
```
Comment (yechs) [inline src/components/views/CourseScheduler/Course.jsx:222]: I think this field may no longer be formatted at all... Might just be a UUID.

Here are some examples that may persuade you:

| Semester | Institution | `termId` field | sample link |
|--|--|--|--|
| 2024 Winter | Williams |100082571 | [link](https://www.bkstr.com/williamsstore/course-materials-results?shopBy=course&divisionDisplayName=&departmentDisplayName=PHIL&courseDisplayName=14&sectionDisplayName=01&programId=912&termId=100082571) |
| 2024 Winter | Stanford |100080358 | [link](https://www.bkstr.com/stanfordstore/course-materials-results?shopBy=course&divisionDisplayName=Stanford&departmentDisplayName=PHIL&courseDisplayName=1&sectionDisplayName=01&programId=562&termId=100080358) |
| 2024 Winter | NYU |100082037 | [link](https://www.bkstr.com/nyustore/course-materials-results?shopBy=course&divisionDisplayName=&departmentDisplayName=ANTH-UA&courseDisplayName=319&sectionDisplayName=001&programId=4582&termId=100082037) |
| 2024 Spring | Williams |100082572 | [link](https://www.bkstr.com/williamsstore/course-materials-results?shopBy=course&divisionDisplayName=&programId=912&departmentDisplayName=PHIL&courseDisplayName=225&sectionDisplayName=T1&termId=100082572) |
| 2024 Spring | Pomona |100081752 | [link](https://www.bkstr.com/claremontstore/course-materials-results?shopBy=course&divisionDisplayName=POMONA%20COLLEGE&departmentDisplayName=PHIL&courseDisplayName=030&sectionDisplayName=01&programId=1550&termId=100081752) |

IMO there are no rules to them at all...

File: src/lib/scheduler.js [hunk 4]
```diff
@@ -119,10 +119,11 @@ const getDefaultSemesterIndex = () => {
   let result = 0;
 
   const now = new Date();
+  console.log(now);
```
Comment (yechs) [inline src/lib/scheduler.js:122]: delete this (debug code) before merging

**Review by yechs — CHANGES_REQUESTED** [review]
TL;DR Very good work.

Action Items:
- [x] I proposed a different logic in scheduler.js:L126
- [x] Remove debug codes before merging
- [x] Maybe change `semesterInNumber` to a constant instead. I don't think it can be calculated
- [ ] (I'll do this) Contact whoever is in charge of the official catalog regarding the link issue. They have this as well.

**Review by AndrewMuh — CHANGES_REQUESTED** [review]

File: src/components/views/CourseScheduler/Course.jsx (hunk 3 above)
Comment (charlietharas) [reply src/components/views/CourseScheduler/Course.jsx:222]: Agreed, definitely a UUID. I'll delete the `switch` because nobody is getting textbooks for winter study anyway, so then currently this only works for spring. I wasn't able to find the Fall UUID either.

**Comment by charlietharas** [comment]
All of your comments should be addressed

**Review by AndrewMuh — APPROVED** [review]
LGTM

File: src/components/views/CourseScheduler/Course.jsx [hunk 5]
```diff
@@ -207,18 +207,12 @@ const Course = ({
...
-    switch (course.semester) {
-      case "Fall":
-        return yearVal - 2; // this does not currently work!
-      case "Winter":
-        return yearVal;
-      default:
-        return yearVal + 1;
-    }
+    const yearVal = 100082572;
+    return yearVal;
```
Comment (yechs) [inline src/components/views/CourseScheduler/Course.jsx:215]: honestly, I think this should best be a config field.
It saves us from having to edit the code.

File: src/components/views/CourseScheduler/Course.jsx [hunk 6]
```diff
@@ -207,18 +207,12 @@ const Course = ({
...
   };
```
Comment (yechs) [inline src/components/views/CourseScheduler/Course.jsx:216]: honestly, I think this should best be a config field.
It saves us from having to edit the code.

File: src/components/views/CourseScheduler/Course.jsx [hunk 7]
```diff
@@ -207,18 +207,12 @@ const Course = ({
...
-    const yearVal = 100080547 + course.year;
-    switch (course.semester) {
-      case "Fall":
-        return yearVal - 2; // this does not currently work!
-      case "Winter":
-        return yearVal;
-      default:
-        return yearVal + 1;
-    }
+    const yearVal = 100082572;
```
Comment (yechs) [inline src/components/views/CourseScheduler/Course.jsx:214]: Should we put this in [`src/constants/constants.js`](https://github.com/WilliamsStudentsOnline/wso-react/blob/1b6903d/src/constants/constants.js)?

It would make later management a lot easier.

I'm even thinking of making it an API endpoint and do some scheduled crawling work in the backend that fetches the correct semesterID.

File: src/lib/scheduler.js [hunk 8]
```diff
@@ -121,14 +121,14 @@ const getDefaultSemesterIndex = () => {
   const now = new Date();
   // Check if Winter (Period 3, above)
   if (
-    new Date(DATES.PREREG.WINTER) < now &&
-    now < addDays(new Date(DATES.Winter.END), -7)
+    new Date(gcalFormattedDate(DATES.PREREG.WINTER)) < now &&
+    now < new Date(gcalFormattedDate(DATES.Winter.START))
```
Comment (yechs) [inline src/lib/scheduler.js:125]: Good work. I never noticed that this function exists...

One slight thing. Maybe rename this function to something like `RFC3339FormattedDate` or `JavaScriptFormattedDate`? Since the function returns RFC3339-formatted date, not ICS/gCal formatted dates...

**Comment by yechs** [comment]
BTW can this be fixed maybe this Tuesday? I'd like to merge and deploy it soon.

**Comment by charlietharas** [comment]
New comments addressed

**Review by yechs — APPROVED** [review]
Thanks! Merge this anytime.
